### ISSUE TITLE
Generic class attributes cause type of class to change

### ISSUE THREAD

**Issue by philj56** [astral-sh/ty#2955]
### Summary

This is quite contrived, but if a generic parameter in a class constructor is assigned to a public attribute, in certain situations the type of the class can change and any un-typehinted attributes change their type to `object`. See below:
```python
from typing import reveal_type

class Y[T: int]:
    def __init__(self, val: T):
        self._a: T = val  # private attribute
        self.b = 3

class Z[T: int]:
    def __init__(self, val: T):
        self.a: T = val  # public attribute
        self.b = 3

def func[T](param_type: type[T]):
    param = param_type()
    if isinstance(param, Y):
        reveal_type(param)  # `T@func & Y[object]`
        reveal_type(param.b)  # `Unknown | Literal[3]`
    if isinstance(param, Z):
        reveal_type(param)  # `T@func & Top[Z[Unknown]]`
        reveal_type(param.b)  # `object`

```

I assume the change to `object` comes from the `Top` stuff, but I'm not sure why changing the visibility of an attribute should change anything? (Or why the type of `Top[Z[Unknown]]` isn't just `Z[object]`)

### Version

ty 0.0.20 (2bd025243 2026-03-02)

**Comment by carljm** [comment]
Thanks for the report!

For PEP 695 typevars, variance is inferred. If a mutable attribute is typed with a typevar, the mutability makes that typevar invariant: `Z[bool]` cannot be a subtype of `Z[int]`, since the latter supports `.a = 1` and the former does not. Private attributes are excluded from variance inference, since they should not be externally accessed at all, let alone externally mutated. So far we are discussing behaviors that all type checkers share. (Though in ty we still need #847 also in order to make this exclusion sound.)

Where ty differs from other type checkers is that they will take `isinstance(param, Y)` to narrow to `Y[Unknown]`, silently introducing an unsound gradual type. In ty we narrow to a fully static type, the union of all possible specializations of `Y`, also called the "top materialization" of `Y[Unknown]` (where "materialization" means to substitute the "Unknown" with some static type).

For `Y`, which is covariant in `T`, the top materialization is simply `Y[object]` (since that is a super-type of all possible specializations of `Y`).

For `Z`, which is invariant in `T`, `Z[object]` is not a super-type of other specializations of `Z`, which is why we see the dedicated syntax `Top[Z[Unknown]]` for the top materialization -- it represents the (infinite) union of all possible specializations of `Z`.

So far, so good - but here we reach the bug in ty.

Since `b` is an implicit attribute with no type declaration, we currently infer its type as `Unknown | Literal[3]`. (We plan to change this so we infer `int` instead, see #1240).

When we access an attribute on a top materialization, we currently top materialize the attribute type. The top materialization of `Unknown` is `object`, therefore the top materialization of `Unknown | Literal[3]` is also `object`. That's why the type of `param.b` gets inferred as `object`.

Here's a stripped-down repro of just the buggy part:

```py
from typing import reveal_type, Any

class Z[T: int]:
    a: T
    b: Any

def func[T](param: object):
    if isinstance(param, Z):
        reveal_type(param)
        reveal_type(param.a)  # `object` is correct here
        reveal_type(param.b)  # should be `Any`, we say `object`
```

We should not top-materialize the type of all attributes of a top-materialized type. We should only top-materialize the typevar.

**Comment by carljm** [comment]
The immediate workaround for your case (if you need one) is to annotate the `b` attribute with an explicit declared type (e.g. `int`) -- as long as it's not a gradual type (no `Any`), that will avoid this bug.

**Comment by philj56** [comment]
Wow, nice write-up! Makes sense to me, thanks for the fast response.
